tools/catalog_builder/catalog_builder.py

```diff
@@ -0,0 +1,326 @@
+# for the preprocessor and the MDTF-diagnostics framework
+
+import click
+import intake
+import os
+import pathlib
+import sys
+import time
+import traceback
+import typing
+import xarray as xr
+import yaml
+from datetime import datetime, timedelta
+from ecgtools import Builder
+from ecgtools.builder import INVALID_ASSET, TRACEBACK
+from ecgtools.parsers import parse_cmip6
+from ecgtools.parsers.cesm import parse_cesm_timeseries
+
+
+# The ClassMaker is cribbed from SO
+# https://stackoverflow.com/questions/1176136/convert-string-to-python-class-object
+# Classmaker and the @catalog_class.maker decorator allow class instantiation from
+# strings. The main block can simply call the desired class using the convention
+# argument instead of messy if/then/else blocks. Yes, both work, but I wanted
+# to try something that, if it is not more "Pythonic", is more extensible
+
+class ClassMaker:
+    def __init__(self):
+        self.classes = {}
+
+    def add_class(self, c):
+        self.classes[c.__name__] = c
+
+    # define the class decorator to return the class passed
+    def maker(self, c):
+        self.add_class(c)
+        return c
+
+    def __getitem__(self, n):
+        return self.classes[n]
+
+
+# instantiate the class maker
+catalog_class = ClassMaker()
+
+
+# custom parser for pp data stored on GFDL archive filesystem
+def parse_gfdl_pp_ts(file_name: str):
+    # files = sorted(glob.glob(os.path.join(file_name,'*.nc')))  # debug comment when ready to run
+    # file = pathlib.Path(files[0])  # debug comment when ready to run
+    file = pathlib.Path(file_name)  # uncomment when ready to run
+
+    try:
+        # isolate file from rest of path
+        stem = file.stem
+        # split the file name into components based on _
+        split = stem.split('.')
+        realm = split[0]
+        time_range = split[1]
+        variable_id = split[2]
+        source_type = file.parts[3]
+        member_id = file.parts[4]
+        experiment_id = file.parts[5]
+        source_id = file.parts[6]
+        freq = file.parts[10]
+        chunk_freq = file.parts[11]
+        variant_label = ""
+        grid_label = ""
+        table_id = ""
+        if 'mon' in freq.lower():
+            output_frequency = 'mon'
+        elif 'day' in freq.lower():
+            output_frequency = 'day'
+        elif '6hr' in freq.lower():
+            output_frequency = '6hr'
+        elif 'subhr' in freq.lower():
+            output_frequency = 'subhr'
+
+        # call to xr.open_dataset required by ecgtoos.builder.Builder
+        with xr.open_dataset(file, chunks={}, decode_times=False) as ds:
+            # variable_list = [var for var in ds if 'standard_name' in ds[var].attrs or 'long_name' in ds[var].attrs]
+            # assert(variable_id in variable_list), \
+            # "Did not find variable with standard_name or long_name {variable_id}" \
+            # "in {file}"
+            info = {
```

Comment:
info could come from a separately maintained configuration. Again, not necessary for this PR. However, once we arrive at the template (headers and aggregation columns) work for NCAR, GFDL and CMIP, we could "recommend" on a specific configuration.